### ISSUE TITLE
Train weights to outcome nodes independently

### ISSUE THREAD

**Issue by kuchenrolle** [fchollet/keras#5159]
I have a single-layer dense NN with a linear activation function that I train one data point at a time, with discrete inputs and outputs (so vectors of zeros and one, not a regression case or anything). I'm trying to come up with a loss function that results in weight updates that change the weights to a node in the output layer **only with respect to the error in that node**, but not wrt the error in the other nodes. In other words I want the weights to be the same as if I trained a separate binary network for each outcome. So the change in weight from input_x to output_y should be: 

![image](http://i.imgur.com/O4gdkHf.png)

In case the image doesn't work, this is the latex formula:
`w_{ij_{t+1}} = w_{ij_t} + a_i * lr *(Y_{j} - \sum_{i}^N a_i * w_{ij_{t}})`

I tried the summed squared error, but that doesn't show the desired behaviour. Here's the simplified code I'm using with the weights that I get, as well as the weights that I should be getting:

```
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.optimizers import SGD
from keras.layers import Dense
from keras import backend as K


def summed_squared_error(y_true, y_pred):
    return K.sum(K.square(y_pred - y_true))/2

def run(cues, outcomes):
    num_cues = cues.shape[1]
    num_outcomes = outcomes.shape[1]
    network = Sequential()
    network.add(Dense(num_outcomes, input_dim = num_cues, init = "zero", activation = "linear", bias = False))
    network.compile(loss = summed_squared_error, optimizer = SGD(lr = 0.01))
    network.fit(cues, outcomes, nb_epoch = 1, batch_size = 1, shuffle = False)
    return network

# Cues    Outcomes    Frequency
# tone    shock   1
# yellow  food    1
# tone_blue   shock   1
first_cues = np.array([[0,1,0],[0,0,1],[1,1,0]])
first_outcomes = np.array([[0,1],[1,0],[0,1]])

# Cues    Outcomes    Frequency
# tone    shock   1
# yellow  food    1
# tone_blue   drink   1
second_cues = np.array([[0,1,0],[0,0,1],[1,1,0]])
second_outcomes = np.array([[0,0,1],[0,1,0],[1,0,1]])


first = run(first_cues, first_outcomes)

# target
# .       0.0099
# .       0.0199
# 0.0100  .

# first.get_weights()
# [[ 0.    ,  0.0099],
# [ 0.    ,  0.0199],
# [ 0.01  ,  0.    ]]

second = run(second_cues, second_outcomes)

# target
# 0.0100      0.       -0.0001
# 0.0100      0.       0.0099
# 0.          0.0100   0.

# second.get_weights()
# [[ 0.01  ,  0.    ,  0.0099],
# [ 0.01  ,  0.    ,  0.0199],
# [ 0.    ,  0.01  ,  0.    ]],
```

Please someone help. :|

**Comment by bstriner** [comment]
That is how NNs always work. The weights for a node in the last layer only depend on the inputs and targets for that node.

However, the hidden layers try to find representations for all the nodes in the last layer.

If you really want k models trained independently with different hidden layers and nothing shared, then just create k different models.

Why are you trying to make independent predictions instead of learning everything at the same time? I can't think of any reason that would be preferred.

Also, why are you initializing to zeros and no bias? That is terrible and will learn badly if it learns anything. Do some reading on initialization and symmetry-breaking. Zero weight initialization and no bias is a useless layer.

Cheers

**Comment by kuchenrolle** [comment]
Hey,

there are no hidden layers in my model. If this is how NNs always work, then why don't I get the expected behaviour?

The reason I want my network to behave this way is that I want to use it as an implementation of a very influential learning model ([Rescorla-Wagner](https://en.wikipedia.org/wiki/Rescorla%E2%80%93Wagner_model)) - so I don't want to fit a model that performs particularly well, but one that learns according to a given rule, which allows me to test the predictions the theory makes.

**Comment by bstriner** [comment]
Could be an interesting project. I was a psych major back in my day. 

Neural networks always use the rule W_{t+1} = W_t - lr * loss. You just need to figure out what loss you want to use. If you really want to try your original equation, rearrange as W = W-lr*loss, then code your loss as a custom loss function.

I can't tell if you just want the output weights to be independent or the entire network to be independent. If you want separate independent networks for binary classification, then do separate binary classification networks. However, that is usually considered silly and inefficient. The power of neural networks is that they try to find hidden representations that help predict the outputs.

Your equation doesn't look like it will learn anything. A loss like mean squared error will learn a function.

Your loss is just sum of the errors. That means if it is off by +100 then off by -100, it's loss is 0. It also means if the network predicts -1 million, it will have hugely negative loss. Basically, a function that minimizes your loss won't learn anything useful.

If you put an abs or a square into your equation, it should work.

Also, why no hidden layers? You're just learning a linear function.

**Comment by kuchenrolle** [comment]
Hey,

again, there is no hidden layer, because that is not what the Rescorla-Wagner model does. It's a linear classifier. It's meant to model animal and human learning, not to find the best solution to a given problem.

In this case, training the weights to each outcome separately should be equivalent to training separate networks per outcome, because the RW-rule updates weights only with respect to a given outcome, ignoring how wrong it was about all the other outcomes on that trial.

I don't understand why the equation looks like it doesn't learn anything. The target outcome (Y_j) and the input (a_i) can be 1 or 0 and the weights (w_ij) can be any real-valued number. So if the sum of all weights to the given outcome is smaller than the the target, all those weights will get increased by the difference, scaled by the learning rate. And if the sum is bigger than the target, then the weights will get decreased analogously.

The RW-rule is not a loss function, though. It is specified for the weights to a given outcome. The loss function, on the other hand, is specified over the activation in all outcomes and the their target activations. So I need a loss function, whose derivative wrt to the individual weight will match the equation above. I tried the summed squared error (in the code, above), for which I thought the result would be exactly what I need, but apparently it isn't.

So that is what I'm asking for. What loss function will result in a derivative wrt to the weights that matches the RW learning rule.

**Comment by bstriner** [comment]
At that point you're probably better off just writing your own updates, since you're not trying to use Adam or anything else fancy. Although it would be interesting to see if you could integrate RW or something.

```python
d = Dense(....)
W_t = somefunctionof(d.W)
model.train_function = K.function(inputs, outputs, updates=[(d.W, d.W_t)])
```

If you overwrite the train_function, model.fit should still work correctly if you make sure your function exactly matches the inputs and outputs of the base model.

Alternatively, you could subclass model and override _make_train_function and put your code there.

Cheers

**Comment by bstriner** [comment]
Thinking about it some more, the best way is to write a custom optimizer. The custom optimizer just needs to take in the parameters and the inputs, etc. and return the updates, and keras will build that into the training function for you.
